#python运行os.system(cmd)的退出问题
###发表时间：2020-02-26
###分类：python,shell,经验
###iteye原始地址：<a href="https://kanpiaoxue.iteye.com/admin/blogs/2512631" target="_blank">https://kanpiaoxue.iteye.com/admin/blogs/2512631</a>

---

<div class="iteye-blog-content-contain" style="font-size: 14px;"> 
 <p>&nbsp;</p> 
 <p>发现之前有人写了一个python的脚本使用os.system(cmd)调用shell。 shell异常结束了，os.system(cmd)也返回非0 ，但是python的退出码是0。</p> 
 <p>示例代码如下：</p> 
 <pre name="code" class="shell">#run.sh的内容
#这个命令一定会失败的，因为jar的参数不对。
hadoop jar hello 
</pre> 
 <p>&nbsp;下面是调用这个run.sh的python脚本</p> 
 <pre name="code" class="java">#hello.py的内容
import os
os.system('sh run.sh')</pre> 
 <p>&nbsp;</p> 
 <p>期望的结果：</p> 
 <p style="padding-left: 30px;">运行hello.py ，python内部调用run.sh。由于run.sh里面的hadoop命令失败，sh run.sh会返回非0的错误退出码;然后hello.py会使用这个非零的错误退出码结束程序。</p> 
 <p>实际的结果：</p> 
 <p style="padding-left: 30px;">运行hello.py ，python内部调用run.sh。由于run.sh里面的hadoop命令失败，sh run.sh会返回非0的错误退出码;但是运行python程序hello.py的退出码是0。和预期不符。</p> 
 <p>问题原因：</p> 
 <p>原因很简单。看下面的python代码就明白了。</p> 
 <pre name="code" class="java">#hello.py的内容
import os
import sys

exit_code = os.system('sh run.sh')
print 'exit_code:%s' % exit_code 
# 这里返回的是非 0 ，因为run.sh运行出现异常返回的就是非0。
# 加入如下代码，让当前python脚本也使用非0退出
if 0 != exit_code:
  sys.exit(1)
</pre> 
 <p>如上面代码。&nbsp;&nbsp;exit_code = os.system('sh run.sh') 里面的&nbsp;exit_code 因为run.sh 返回非0，但是实际上python的这个函数os.system('sh run.sh')运行是成功的，只不过函数的返回值是非0.这里要区分python脚本运行的退出码和os.system(cmd)的返回值不是一回事，需要根据os.system('sh run.sh')的函数返回值来判断如何让python退出。</p> 
 <p>&nbsp;</p> 
 <p>&nbsp;</p> 
</div>